src/pools/D3MAaveV3TypePool.sol

```diff
@@ -0,0 +1,292 @@
+            z = x != 0 ? ((x - 1) / y) + 1 : 0;
+        }
+    }
+
+    // --- Admin ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        require(vat.live() == 1, "D3MAaveV3TypePool/no-file-during-shutdown");
+        if (what == "hub") {
+            vat.nope(hub);
+            hub = data;
+            vat.hope(data);
+        } else if (what == "king") king = data;
+        else revert("D3MAaveV3TypePool/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    // Deposits Dai to Aave in exchange for adai which is received by this contract
+    // Aave: https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#deposit
+    function deposit(uint256 wad) external override onlyHub {
+        uint256 scaledPrev = adai.scaledBalanceOf(address(this));
+
+        pool.deposit(address(dai), wad, address(this), 0);
+
+        // Verify the correct amount of adai shows up
+        uint256 interestIndex = pool.getReserveNormalizedIncome(address(dai));
+        uint256 scaledAmount = _rdiv(wad, interestIndex);
+        require(adai.scaledBalanceOf(address(this)) >= (scaledPrev + scaledAmount), "D3MAaveV3TypePool/incorrect-adai-balance-received");
+    }
+
+    // Withdraws Dai from Aave in exchange for adai
+    // Aave: https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#withdraw
+    function withdraw(uint256 wad) external override onlyHub {
+        uint256 prevDai = dai.balanceOf(msg.sender);
+
+        pool.withdraw(address(dai), wad, msg.sender);
+
+        require(dai.balanceOf(msg.sender) == prevDai + wad, "D3MAaveV3TypePool/incorrect-dai-balance-received");
+    }
+
+    function exit(address dst, uint256 wad) external override onlyHub {
+        uint256 exited_ = exited;
+        exited = exited_ + wad;
+        uint256 amt = wad * assetBalance() / (D3mHubLike(hub).end().Art(ilk) - exited_);
+        require(adai.transfer(dst, amt), "D3MAaveV3TypePool/transfer-failed");
+    }
+
+    function quit(address dst) external override auth {
+        require(vat.live() == 1, "D3MAaveV3TypePool/no-quit-during-shutdown");
+        require(adai.transfer(dst, adai.balanceOf(address(this))), "D3MAaveV3TypePool/transfer-failed");
+    }
+
+    function preDebtChange() external override {
+        // Execute a 1 wei flash loan to trigger an index update on everything
+        // If the fee is less than 100% than the fee should always round down to zero
+        // Note: we can ignore indicies being up to date if the supply cap is 0
+        PoolLike.ReserveData memory data = pool.getReserveData(address(dai));
+        if (
+            // The next two lines can be slightly optimized, but it's easier to read like this
+            (data.configuration & ~FLASHLOAN_ENABLED_MASK) != 0 &&
```

Comment:
What would be the implications if flash loans are disabled?

Reply:
Also why do we need to update the indexes here vs the v2 d3m?

Reply:
I added success and failure tests to cover this case.

Reply:
I think I like your suggestion to drop the supply cap support, which would narrow the problem space.
We can have a supply cap check in active() and unwind if there is one.
It can be tackled again when AaveV3 pools with supply caps are supported.

I'm not 100% sure but I don't think it would require changing to a new name (e.g `SparkType`), since once supply caps are supported it will just be considered an upgrade and will work also for Spark.

Reply:
Yes I consider the same than Chris, if AaveV3 official implementation uses the cap, and Spark not, I tend to rename it to `SparkTypePool`. Now if AaveV3 doesn't currently use caps either and will start using it in the future, maybe we could name it a `AaveV3TypePoolNoCap`.

Reply:
I really think it will be more confusing to mix instance vs code for naming. The codebase is still Aave V3 even if for Spark Lend the caps are turned off. What if someone else deploys another instance of Aave V3 with the caps turned off? Does it make sense to use a `SparkTypePool` adapter? There's no relation there, but it's the same code.

I think as a compromise I'd be okay with adding the "NoCap" to the end so it is `AaveV3NoSupplyCapTypePool` and splitting out the cap support into a new PR (which doesn't have to be looked at until it's needed). How does that sound?

Reply:
is it a code change what you are doing in Spark for removing the caps or just setting a value to max?

Reply:
> is it a code change what you are doing in Spark for removing the caps or just setting a value to max?

There is no code changes between the official Aave V3 instance and Spark Lend. The only difference is configuration.

Reply:
Ok then I think you are probably right and the correct name should be like `AaveV3NoSupplyCapTypePool`.

Reply:
Seems like you all are in agreement, but I'll just weigh in here and to say again that this feels to me like we're getting the worst of both specificity and abstraction.  we have a very specific "abstract" implementation for a certain configuration of code and i'm not really sure how its better than just calling the code after the instance its pointed to.   If we follow this logic, i'm worried we'll end up with something like `AaveV3NoSupplyCapNonRebasingStableRateDaiOnlyTypePool` in the future which just seems like the wrong place to end up. 

Plus if we go this route and actually do find two use cases for our "abstract" `AaveV3NoCapTypePool` and then some other configuration changes between those two we'll be right back to this debate of adding contextual/complex code to the pool or separating them again to a configuration abstraction (see naming above).

I would vote for 
1) leave this abstract `AaveV3TypePool` with the code so it works against the Aave code base regardless of how it is configured
or
2) build a `SparkPool` that works for the D3M we are building that governance has decided we should build and then have an AaveV3Pool or FluxPool or WhateverSubDAOPool in the future when we know what that looks like

With a preference for 2 because 1 feels like we're trying to optimize for a future we don't really know yet and paying a complexity price today for unknown benefit...

Reply:
> i'm worried we'll end up with something like `AaveV3NoSupplyCapNonRebasingStableRateDaiOnlyTypePool` in the future which just seems like the wrong place to end up.

Yes this situation is not good either. To be honest I don't think there is a perfect solution, as we discussed in the very beginning. All of them will have pros and cons. I still feel better the way it was recently pushed because describes more the code than the project where it will used, but yeah certainly if we end up in a situation as you put as example that will be very bad as well.